src/frontend/src/optimizer/plan_node/batch_log_seq_scan.rs

```diff
@@ -0,0 +1,163 @@
+            match self.core.distribution_key() {
+                None => Distribution::SomeShard,
+                Some(distribution_key) => {
+                    if distribution_key.is_empty() {
+                        Distribution::Single
+                    } else {
+                        Distribution::UpstreamHashShard(
+                            distribution_key,
+                            self.core.table_desc.table_id,
+                        )
+                    }
+                }
+            },
+        )
+    }
+
+    /// Get a reference to the batch seq scan's logical.
+    #[must_use]
+    pub fn core(&self) -> &generic::LogScan {
+        &self.core
+    }
+}
+
+impl_plan_tree_node_for_leaf! { BatchLogSeqScan }
+
+impl Distill for BatchLogSeqScan {
+    fn distill<'a>(&self) -> XmlNode<'a> {
+        let verbose = self.base.ctx().is_explain_verbose();
+        let mut vec = Vec::with_capacity(4);
+        vec.push(("table", Pretty::from(self.core.table_name.clone())));
+        vec.push(("columns", self.core.columns_pretty(verbose)));
+
+        if verbose {
+            let dist = Pretty::display(&DistributionDisplay {
+                distribution: self.distribution(),
+                input_schema: self.base.schema(),
+            });
+            vec.push(("distribution", dist));
+        }
+
+        childless_record("BatchScan", vec)
+    }
+}
+
+impl ToDistributedBatch for BatchLogSeqScan {
+    fn to_distributed(&self) -> Result<PlanRef> {
+        Ok(self.clone_with_dist().into())
+    }
+}
+
+impl TryToBatchPb for BatchLogSeqScan {
+    fn try_to_batch_prost_body(&self) -> SchedulerResult<NodeBody> {
+        Ok(NodeBody::LogRowSeqScan(LogRowSeqScanNode {
+            table_desc: Some(self.core.table_desc.try_to_protobuf()?),
+            column_ids: self
+                .core
+                .output_column_ids_to_batch()
+                .iter()
+                .map(ColumnId::get_id)
+                .collect(),
+            vnode_bitmap: None,
+            old_epoch: Some(BatchQueryEpoch {
+                epoch: Some(risingwave_pb::common::batch_query_epoch::Epoch::Committed(
+                    self.core.old_epoch,
+                )),
+            }),
+            new_epoch: Some(BatchQueryEpoch {
+                epoch: Some(risingwave_pb::common::batch_query_epoch::Epoch::Committed(
+                    self.core.new_epoch,
+                )),
+            }),
+        }))
+    }
+}
+
+impl ToLocalBatch for BatchLogSeqScan {
+    fn to_local(&self) -> Result<PlanRef> {
+        let dist = if let Some(distribution_key) = self.core.distribution_key()
+            && !distribution_key.is_empty()
+        {
+            Distribution::UpstreamHashShard(distribution_key, self.core.table_desc.table_id)
+        } else {
+            Distribution::SomeShard
+        };
+        Ok(Self::new_inner(self.core.clone(), dist).into())
+    }
+}
+
+impl ExprRewritable for BatchLogSeqScan {
```

Comment:
We can return false, and then use the default rewrite_exprs implementation defined in the trait. It seems no be no-op in `core.rewrite_exprs(r);`.